common-rest/src/main/java/org/uniprot/api/rest/controller/BasicSearchController.java

```diff
@@ -114,26 +115,35 @@ protected BasicSearchController(
         } else {
             context.setEntities(result);
         }
-
-        return getResponseBodyEmitterResponseEntity(request, context);
+        return getDeferredResultResponseEntity(request, context);
     }
 
-    protected ResponseEntity<ResponseBodyEmitter> getResponseBodyEmitterResponseEntity(
-            HttpServletRequest request, MessageConverterContext<T> context) {
-        ResponseBodyEmitter emitter = new ResponseBodyEmitter();
+    protected DeferredResult<ResponseEntity<MessageConverterContext<T>>>
+            getDeferredResultResponseEntity(
+                    HttpServletRequest request, MessageConverterContext<T> context) {
+
+        // timeout in millis
+        Long timeoutInMillis = Long.valueOf(downloadTaskExecutor.getKeepAliveSeconds()) * 1000;
+
+        DeferredResult<ResponseEntity<MessageConverterContext<T>>> deferredResult =
+                new DeferredResult<>(timeoutInMillis);
+
+        deferredResult.onTimeout(() -> log.error("Request timeout occurred."));
+
         downloadTaskExecutor.execute(
                 () -> {
                     try {
-                        emitter.send(context, context.getContentType());
-                    } catch (IOException e) {
-                        emitter.completeWithError(e);
+                        deferredResult.setResult(
+                                ResponseEntity.ok()
+                                        .headers(createHttpDownloadHeader(context, request))
+                                        .body(context));
+                    } catch (Exception e) {
+                        log.error("Error occurred during processing.");
+                        deferredResult.setErrorResult(e);
```

Comment:
Wondering whether we should push this logic down into the service layer, to keep the controller as thin as possible.

Reply:
I agree with this. But it will require a lot of changes. We depend on many fields of BaseController which comes from controller. Can we do this refactoring as a part of another ticket?

Reply:
I think this logic is related with how we return the data, and handle any error during it, which,  I think it is controller related logic. Please, correct me if I am wrong.
I think that in the service layer we would put more Business Logic.